File: 0030-substring-with-concatenation-of-all-words/0030-substring-with-concatenation-of-all-words.py
class Solution:
    def findSubstring(self, s: str, words: List[str]) -> List[int]:
        n = len(s)
        wordLen = len(words[0])
        wordCount = len(words)
        ans = []

        totalLength = wordLen * wordCount

        if totalLength > n :
            return ans
        
        required = {}

        for i in range(wordCount):
            if words[i] not in required:
                required[words[i]] = 0

            required[words[i]] += 1
        
        for offset in range(wordLen):

            l =offset
            r =offset

            current = {}
            count =0

            while r+wordLen <=n:
                word = s[r:r+wordLen]
                r+= wordLen

                if word not in required:
                    current.clear()
                    count =0
                    l=r

                else :

                    if word not in current:
                        current[word] =0
                    current[word]+=1
                    count +=1

                    while current[word] > required[word]:
                        leftWord = s[l:l+wordLen]
                        current[leftWord]-=1
                        count-=1
                        l+=wordLen
                    
                    if count == wordCount:

                        ans.append(l)
                        leftWord = s[l:l+wordLen]
                        current[leftWord]-=1
                        count -=1
                        l+=wordLen
        
        return ans 

File: 0030-substring-with-concatenation-of-all-words/0030-substring-with-concatenation-of-all-words.cpp
class Solution {
public:
    vector<int> findSubstring(string s, vector<string>& words) {

        vector<int> ans;

        int m = s.length();
        int wordLen = words[0].length();
        int wordCount = words.size();

        int totalLen = wordLen * wordCount;

        if (totalLen > m)
            return ans;

        unordered_map<string, int> required;

        for (string word : words) {
            required[word]++;
        }

        // We need wordLen different starting offsets
        for (int offset = 0; offset < wordLen; offset++) {

            int l = offset;
            int r = offset;

            unordered_map<string, int> current;
            int count = 0;

            while (r + wordLen <= m) {

                // Take one word from s
                string word = s.substr(r, wordLen);
                r += wordLen;

                // Word is not required
                if (required.find(word) == required.end()) {

                    current.clear();
                    count = 0;
                    l = r;
                }
                else {

                    current[word]++;
                    count++;

                    // Too many occurrences of this word
                    while (current[word] > required[word]) {

                        string leftWord = s.substr(l, wordLen);

                        current[leftWord]--;
                        count--;

                        l += wordLen;
                    }

                    // We have exactly all required words
                    if (count == wordCount) {

                        ans.push_back(l);

                        // Move left to look for next window
                        string leftWord = s.substr(l, wordLen);

                        current[leftWord]--;
                        count--;

                        l += wordLen;
                    }
                }
            }
        }

        return ans;
    }
};